README.md
# taiwan-esun-credit-card


```bash
composer require june1hub/taiwan-esun-credit-card
```

CHANGELOG.md
## Change Log

### 0.0.1 - 2021-08-06
- 各種初始化


src/EsunCreditCard.php
<?php

namespace June1hub\TaiwanEsunCreditCard;

use Illuminate\Support\Facades\View;
use June1hub\TaiwanEsunCreditCard\Traits\ResponseTrait;

class EsunCreditCard
{
    use ResponseTrait;
    
    /**
     * 付款
     * 
     * @param params.ONO *(string) : 訂單編號(不可重複)
     * @param params.TA  *(integer): 金額
     * @param params.U   *(url)    : 回傳網址
     * @param params.IC   (integer): 分期 3|6
     * @param params.BPF  (string) : 銀行紅利折抵 Y|N
     */
    public static function pay($params)
    {
        $config = config('esun.creditCard');

        // 網址
        $apiUrl = config('app.env') == 'production' ? $config['apiUrl']['live'] : $config['apiUrl']['test'];
        
        // 特店代碼
        $params['MID'] = $config['MID'];
        // 終端機代碼
        $params['TID'] = $config['TID'];

        // 參數整理
        $data = json_encode($params);
        $mac  = hash('sha256', $data . $config['mackey']);
        $ksn  = $config['ksn'];
        $postData = [
            'data' => $data,
            'mac' => $mac,
            'ksn' => $ksn
        ];

        // i don't know why this is not working
        // return view('esun::send', compact('apiUrl', 'postData'));

        // use this temporarily
        echo '
            <!doctype html>
            <html lang="en">
                <head>
                    <meta charset="UTF-8">
                    <meta http-equiv="X-UA-Compatible" content="IE=edge" />
                    <meta name="viewport" content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=0">
                    <title>Redirect to Pay...</title>
                </head>
                <body>
                    <form action="'.$apiUrl.'" id="pay-form" method="POST">';
                        foreach ($postData as $key => $value) {
                            $value = $key == 'data' ? htmlspecialchars($value): $value;
                            echo '<input type="hidden" name="'.$key.'" value="'.$value.'">';
                        }                        
        echo '                        
                    </form>
                </body>
                <script>
                   document.getElementById("pay-form").submit();
                </script>
            </html>
        ';
        
        
        // View::make('EsunCreditCard::send')
        //     ->with('apiUrl', $apiUrl)
        //     ->with('postData', $postData);
    }

    public static function parseResponse($request)
    {
        $data = $request->DATA;

        // 整理
        $newRequest = [];
        $receive_data_temp_1 = explode(',', $data);
        foreach ($receive_data_temp_1 as $key => $value) {
            $receive_data_temp_2 = explode('=', $value);
            $newRequest[$receive_data_temp_2[0]] = $receive_data_temp_2[1];
        }

        if (!isset($request['MACD']) || !isset($newRequest['RC'])) {
            return self::getResponseIsError();
        }

        return [
            'status' => self::getResponseIsSuccess($newRequest['RC']),
            'message' => self::getResponseMessage($newRequest['RC'])
        ];
    }

    

}

src/EsunCreditCardServiceProvider.php
<?php

namespace June1hub\TaiwanEsunCreditCard;

use Illuminate\Support\ServiceProvider;

class EsunCreditCardServiceProvider extends ServiceProvider
{
    public function boot()
    {
        if ($this->app->runningInConsole()) {
            $this->registerConfigs();
        }
        $this->registerResources();
        
        // $this->loadRoutesFrom(__DIR__.'/../routes/routes.php');
    }

    protected function registerConfigs()
    {
        $this->publishes([
            __DIR__ . '/../config/esun.php' => config_path('esun.php')
        ], 'creditcard');
    }

    protected function registerResources()
    {        
        // $this->loadViewsFrom(__DIR__ . '/../resources/views', 'esun');

        $this->loadViewsFrom(__DIR__.'/../resources/views', 'esun');

        // $this->publishes([
        //     __DIR__.'/../resources/views' => resource_path('views/vendor/esun/creditCard'),
        // ]);
    }

}

config/esun.php
<?php
return [
    'creditCard' => [

        'apiUrl' => [
            'live' => 'https://acq.esunbank.com.tw/ACQTrans/esuncard/txnf014s',
            'test' => 'https://acqtest.esunbank.com.tw/ACQTrans/esuncard/txnf014s',
        ],
        
        // mackey
        'mackey' => '',
        
        // 特店代碼
        'MID' => '',

        // 終端機代號 
        'TID' => '',
        
        // MacKey 序號，固定1
        'ksn' => 1,

        'payStatusCode' => [
            '00' => '核准(交易成功)',
            '01' => '請查詢發卡銀行',
            '02' => '請查詢發卡銀行',
            '05' => '請查詢發卡銀行',
            '06' => '拒絕交易',
            '07' => '拒絕交易',
            '08' => '拒絕交易',
            '09' => '拒絕交易',
            "10" => "授權中",
            "11" => "授權失敗",
            '12' => '拒絕交易',
            '13' => '拒絕交易',
            '14' => '卡號錯誤',
            "40" => "授權中",
            "41" => "授權取消",
            "49" => "授權取消",
            "50" => "交易中",
            '51' => '請查詢發卡銀行',
            '54' => '卡片過期',
            '57' => '請查詢發卡銀行',
            "59" => "核准(交易成功)",
            "60" => "退貨中",
            "61" => "退貨失敗",
            '62' => '尚未開卡',
            "69" => "退貨成功",
            '87' => '拒絕交易',
            'L1' => '產品代號錯誤',
            'L2' => '期數錯誤',
            'L3' => '不支援分期',
            'L4' => '產品代碼過期',
            'L5' => '金額無效',
            'L6' => '不支援分期',
            'L7' => '非限定卡別交易',
            'XA' => '紅利自付額有誤',
            'XB' => '紅利商品數量有誤',
            'XC' => '紅利商品數量超過可折抵上限',
            'XD' => '紅利商品折抵點數超過最高折',
            'XE' => '紅利商品傳入之固定點數有誤',
            'XF' => '紅利折抵金額超過消費金額',
            'X1' => '不允許使用紅利折抵現金功能',
            'X2' => '點數未達可折抵點數下限',
            'X3' => '他行卡不支援紅利折抵',
            'X4' => '此活動已逾期',
            'X5' => '金額未超過限額不允許使用',
            'X6' => '特店不允許紅利交易',
            'X7' => '點數不足',
            'X8' => '非正卡持卡人',
            'X9' => '紅利商品編號有誤或空白',
            'ET' => '銀聯卡自訂回覆碼',
            'UE' => '請聯繫收單銀行',
            'TE' => 'Token 格式錯誤',
            'TI' => '無效 Token',
            'G0' => '系統功能有誤',
            'G1' => '交易逾時',
            'G2' => '資料格式錯誤',
            'G5' => '連線IP不合',
            'G6' => '訂單編號重複',
            'G7' => '使用未定義之紅利點數進行交易',
            'G8' => '押碼錯誤',
            'G9' => 'Session 檢查有誤',
            'GA' => '無效的持卡人資料',
            'GB' => '不允許執行授權取消交易',
            'GC' => '退貨期限逾期',
            'GD' => '玉山查無訂單編號',
            'GE' => '查無交易明細',
            'GF' => '交易資料狀態不符',
            'GG' => '交易失',
            'GH' => '訂單編號重複送出交易',
            'GI' => '銀行紅利狀態不符',
            'GJ' => '出團日期不合法',
            'GK' => '延後出團天數超過限定天數',
            'GL' => '非限定特店，不可使用「玉山卡」參數',
            'GM' => '限定特店，必須傳送「玉山卡」參數',
            'GN' => '該卡號非玉山卡所屬',
            'GP' => '銀行紅利與分期只能二選一',
            'GR' => '使用者取消刷卡頁面',
            'GS' => '系統暫停服務',
            'GT' => '交易時間逾時',
            'GU' => '預先授權重覆交易',
            'GV' => '無預先授權成功交易紀錄',
            'GW' => '無預先授權交易紀錄',
            'GX' => '3D 交易異常',
            'GY' => '3D 交易異常',
            'GZ' => '3D 交易異常',
            'V0' => '3D 驗證失敗',
            'V1' => '3D 交易異常',
            'V2' => '發卡行(ACS)系統異常',
            'VA' => '取消交易金額有誤',
            'VD' => '取消交易時間不合法(日期)',
            'VT' => '取消交易時間不合法(時間) ',
            'RA' => '退貨交易金額有誤',
            'RD' => '超過退貨交易期限',
            'RP' => '退貨交易不符',
            'N1' => '紅利點數與金額不符',
            'NM' => '未設定特店類型或英文名稱',
            'NR' => '不允許退貨交易',
            'NT' => '未設定為銀聯卡交易特店',
            'NV' => '不允許取消交易',
            'Z0' => 'URL 錯誤 ',
            'Z1' => '不允許銀聯交易 ',
            'Z2' => '無法進行完全 3D 交易 ',
            'QQ' => '不允許 Debit Card 交易 ', 
        ],

        // 付款成功的 code
        'successResponceCode' => [
            '00',
            '59'
        ],
    ],
];
